### PR TITLE
OAuth: Auth should succeed when birthdate and gender are not returned by providers

### PR DESCRIPTION
**Pull request by drryanjames**
ACTUAL:
Facebook auth fails when a user hides their birthdate or gender

EXPECTED:
Auth succeeds when birthdate or gender are hidden.

Change Summary:
- Moved parsing logic to common function
- Added tests for failing scenarios
- Updated documentation

Before these changes, Facebook and Google json were parsed in separate functions. I checked for birthdate and gender properly in Google's parser, but didn't do so in Facebook's (oops). In addition, neither parser was tested because they were hard to test. These changes move parsing logic to a common function which minimizes provider specific code and allows me to test parse logic using a mock. This should also make new providers easier to add in the future.

### CODE REVIEW

File: portal/models/flaskdanceprovider.py [hunk 1]
```diff
@@ -11,21 +11,134 @@ class FlaskDanceProvider:
...
+                # Handle these cases gracefully
+                if not required and key not in value:
+                    return None
+
+                value = value[key]
+
+            return value
+
+        # Attempt to parse the user's json
+        try:
```
Comment (ivan-c) [inline portal/models/flaskdanceprovider.py]: Would it make sense to move this try/except into `get_value_from_json()`, to get more granular exception info?

File: portal/models/flaskdanceprovider.py [hunk 2]
```diff
@@ -11,21 +11,134 @@ class FlaskDanceProvider:
...
+                required=False
+            )
+            user_info.birthdate = get_value_from_json(
+                'birthdate',
+                required=False
+            )
+
+            return user_info
+        except KeyError as err:
+            current_app.logger.debug(
```
Comment (ivan-c) [inline portal/models/flaskdanceprovider.py]: If we're missing a required value, I would probably consider that an error, ie `current_app.logger.error`

File: portal/models/flaskdanceprovider.py [hunk 3]
```diff
@@ -11,21 +11,121 @@ class FlaskDanceProvider:
...
+        return self.parse_json(user_json)
+
+    def parse_json(self, user_json):
+        """parses the user's json and returns it in a standard format
+
+        Providers encode user information in json. This function parses
+        the json and stores values in an instance of FlaskProviderUserInfo
+
+        :param user_json: info about the user encoded in json
+
```
[the remark this answers is not in the document]
Comment (ivan-c) [reply portal/models/flaskdanceprovider.py]: The mystery of the sphinx ;)

**Comment by ivan-c** [comment]
FYI added @pbugni as a reviewer as well

File: portal/models/flaskdanceprovider.py [hunk 4]
```diff
@@ -11,21 +11,134 @@ class FlaskDanceProvider:
...
+        filled with the user's information
+
+        :return FlaskProviderUserInfo with the user's info
+        """
+
+        # Ask the provider for details about the user
+        resp = self.send_get_user_json_request()
+
+        if not resp.ok:
+            current_app.logger.debug(
```
Comment (pbugni) [inline portal/models/flaskdanceprovider.py]: I'm assuming this is an error/unexpected scenario?  Lets use `current_app.logger.error()` if so, so it doesn't go unnoticed.

File: portal/models/flaskdanceprovider.py [hunk 5]
```diff
@@ -11,21 +11,134 @@ class FlaskDanceProvider:
...
+            self.standard_key_to_provider_key_map links standard keys to
+            provider specific keys, including those that are nested, which
+            allows our parsing code to stay as generic as possible.
+
+            :param standard:key: the standard key
+            :param required: is this property required?
+            :return value from the user's json or None
+            """
+
+            # Ge the key used by the provider
```
Comment (pbugni) [inline portal/models/flaskdanceprovider.py]: typo `Ge`

File: portal/models/flaskdanceprovider.py (hunk 2 above)
Comment (pbugni) [reply portal/models/flaskdanceprovider.py]: Agreed

File: portal/models/flaskdanceprovider.py (hunk 1 above)
Comment (drryanjames) [reply portal/models/flaskdanceprovider.py]: I like that idea. Added a trycatch in ```get_value_from_json``` to record a more granular exception, then rethrow.

File: portal/models/flaskdanceprovider.py (hunk 4 above)
Comment (drryanjames) [reply portal/models/flaskdanceprovider.py]: Not sure why I keep making this mistake :open_mouth:

File: portal/models/flaskdanceprovider.py (hunk 5 above)
Comment (drryanjames) [reply portal/models/flaskdanceprovider.py]: I'm going back to elementary school.

**Pull request merged**